src/common/lib/client/realtimechannel.ts

```diff
@@ -430,15 +441,92 @@ class RealtimeChannel extends Channel {
       return;
     }
 
-    this.subscriptions.on(event, listener);
+    // Filtered
+    if (event && typeof event === 'object' && !Array.isArray(event)) {
+      this._subscribeFilter(event, listener);
+    } else {
+      this.subscriptions.on(event, listener);
+    }
 
     return this.attach(callback || noop);
   }
 
+  _subscribeFilter(filter: MessageFilter, listener: MessageListener) {
+    const filteredListener = (m: Message) => {
+      const mapping: { [key in keyof MessageFilter]: any } = {
+        name: m.name,
+        refId: m.extras?.reference?.id,
+        refType: m.extras?.reference?.type,
+        hasRef: !!m.extras?.reference?.id,
+      };
+      // Check if any values are defined in the filter and if they match the value in the message object
+      if (
+        Object.entries(filter).find(([key, value]) =>
+          value !== undefined ? mapping[key as keyof MessageFilter] !== value : false
+        )
+      ) {
+        return;
+      }
+      listener(m);
+    };
+    this._addFilteredSubscription(filter, listener, filteredListener);
+    this.subscriptions.on(filteredListener);
+  }
+
+  // Adds a new filtered subscription
+  _addFilteredSubscription(filter: MessageFilter, realListener: MessageListener, filteredListener: MessageListener) {
+    if (this.filteredSubscriptions.has(realListener)) {
```

Comment:
should this throw an error instead of quietly overwriting the old filter?

Reply:
Overwriting shouldn't happen here, if the filter and listener are both identical to a previous subscription then the new listener is added to the array (`realListenerMap.set(filter, realListenerMap.get(filter).concat(filteredListener)`)